ruby/private/bundle/bundle.bzl

```diff
@@ -1,123 +1,180 @@
-load("//ruby/private:constants.bzl", "RULES_RUBY_WORKSPACE_NAME")
+load(
+    "//ruby/private:constants.bzl",
+    "BUNDLE_ATTRS",
+    "BUNDLE_BINARY",
+    "BUNDLE_BIN_PATH",
+    "BUNDLE_PATH",
+    "RULES_RUBY_WORKSPACE_NAME",
+    "SCRIPT_BUILD_FILE_GENERATOR",
+    "SCRIPT_INSTALL_GEM",
+)
+load("//ruby/private:providers.bzl", "RubyRuntimeContext")
+load("//ruby/private/tools:deprecations.bzl", "deprecated_attribute")
 
-def install_bundler(ctx, interpreter, install_bundler, dest, version):
-    args = [interpreter, install_bundler, version, dest]
-    environment = {"RUBYOPT": "--disable-gems"}
+# Runs bundler with arbitrary arguments
+# eg: run_bundler(runtime_ctx, [ "lock", " --gemfile", "Gemfile.rails5" ])
+def run_bundler(runtime_ctx, bundler_arguments):
+    #print("BUNDLE RUN", bundler_arguments)
 
-    result = ctx.execute(args, environment = environment)
-    if result.return_code:
-        message = "Failed to evaluate ruby snippet with {}: {}".format(
-            interpreter,
-            result.stderr,
-        )
-        fail(message)
+    # Now we are running bundle install
+    args = [
+        runtime_ctx.interpreter,  # ruby
+        "--enable=gems",  # bundler must run with rubygems enabled
+        "-I",
+        ".",
+        "-I",  # Used to tell Ruby where to load the library scripts
+        BUNDLE_PATH,  # Add vendor/bundle to the list of resolvers
+        BUNDLE_BINARY,  # our binary
+    ] + bundler_arguments
 
-def bundle_install_impl(ctx):
-    ctx.symlink(ctx.attr.gemfile, "Gemfile")
-    ctx.symlink(ctx.attr.gemfile_lock, "Gemfile.lock")
-    ctx.symlink(ctx.attr._create_bundle_build_file, "create_bundle_build_file.rb")
-    ctx.symlink(ctx.attr._install_bundler, "install_bundler.rb")
+    # print("Bundler Command:\n\n", args)
 
-    # TODO(kig) Make Gemspec reference from Gemfile actually work
-    if ctx.attr.gemspec:
-        ctx.symlink(ctx.attr.gemspec, ctx.path(ctx.attr.gemspec).basename)
+    return runtime_ctx.ctx.execute(
+        args,
+        quiet = False,
+        environment = runtime_ctx.environment,
+    )
 
-    ruby = ctx.attr.ruby_interpreter
-    interpreter_path = ctx.path(ruby)
+#
+# Sets local bundler config values by calling
+#
+# $ bundle config --local | --global config-option config-value
+#
+# @config_category can be either 'local' or 'global'
+def set_bundler_config(runtime_ctx, config_category = "local"):
+    # Bundler is deprecating various flags in favor of the configuration.
+    # HOWEVER — for reasons I can't explain, Bazel runs "bundle install" *prior*
+    # to setting these flags. So the flags are then useless until we can force the
+    # order and ensure that Bazel first downloads Bundler, then sets config, then
```

Comment:
To force an order of rules you have to chain inputs and outputs. So inputting the output of the bundler rule as input into config should work.

Reply:
Well, this isn't really about ordering of different rules — this is ordering of Starlark function calls within an implementation of a single rule `ruby_bundle`. Both `bundle install` and `bundle config` run inside of the rule's implementation function, so that's exactly what made me puzzled before.